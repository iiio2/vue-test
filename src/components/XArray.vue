<template>
  <div>
    <input
      v-for="(input, index) in internalValue"
      :key="index"
      v-model="internalValue[index]"
    />
  </div>
</template>

<script>
export default {
  props: ["modelValue"],
  data() {
    return {
      internalValue: [],
    };
  },
  watch: {
    // when modelValue is changed
    modelValue: {
      handler(newVal) {
        // todo: fill internalValue from v-model and add an empty element
        this.internalValue = [...newVal, ""];
      },
      deep: true,
    },
    // when internalValue is changed
    internalValue: {
      handler(newVal) {
        // todo: form a result without empty elements
        // todo: emit input event with result
        const result = newVal.filter(Boolean);
        console.log(newVal);
        if (JSON.stringify(newVal) === JSON.stringify(this.modelValue))
          return null;
        console.log("update:modelValue", result);
        this.$emit("update:modelValue", result);
      },
      deep: true,
    },
  },
  // when component is mounted
  mounted() {
    // fill internalValue from v-model and add an empty element

    this.internalValue = [...this.modelValue, ""];
  },
};
</script>
