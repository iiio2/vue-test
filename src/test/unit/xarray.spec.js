import { mount, shallowMount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import XArray from "../../components/XArray.vue";

describe("XArray", () => {
  it("should accept array of string in v-model", () => {
    const wrapper = mount(XArray, {
      props: {
        modelValue: ["one", "two"],
      },
    });
    expect(wrapper.props().modelValue).toStrictEqual(["one", "two"]);
  });

  it("should render each string as separate input", async () => {
    const wrapper = mount(XArray, {
      props: {
        modelValue: ["one", "two"],
      },
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.findAll("input").length).toBe(3);
  });

  it("writes back to the array when input value is changed", async () => {
    const wrapper = mount(XArray, {
      props: {
        modelValue: ["a", "b", "c"],
      },
    });

    await wrapper.vm.$nextTick();

    const input = wrapper.findAll("input").at(1);
    input.setValue("new value");

    expect(input.element.value).toBe("new value");
  });

  it("removes input element from array when value is empty", async () => {
    const wrapper = mount(XArray, {
      props: {
        modelValue: ["one", "two"],
      },
    });

    await wrapper.vm.$nextTick();
    const input = await wrapper.findAll("input").at(0).setValue("");

    expect(input).toBe(undefined);
  });
  it("should be always one extra input shown, so that it is possible to add new values", async () => {
    const wrapper = mount(XArray, {
      props: {
        modelValue: ["one", "two"],
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll("input").length).toBe(3);
  });
  it("example works as expected (compare the snapshot)", async () => {
    const wrapper = mount(XArray, {
      props: {
        modelValue: ["one", "two"],
      },
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
