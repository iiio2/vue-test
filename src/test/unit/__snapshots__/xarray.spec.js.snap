// Vitest Snapshot v1

exports[`testing XArray > example works as expected (compare the snapshot) 1`] = `"<div><input><input><input></div>"`;

exports[`testing XArray > works as expected 1`] = `"<div><input><input><input></div>"`;

exports[`xarray component testing > should be mounted 1`] = `"<div></div>"`;

exports[`xarray component testing > should expect props 1`] = `"<div></div>"`;
